typed-racket-lib/typed-racket/base-env/prims-contract.rkt

```diff
@@ -250,9 +251,17 @@
 ;;  make-predicate
 ;;  cast
 
-;; Helper to construct syntax for contract definitions
+;; Helpers to construct syntax for contract definitions
+;; make-contract-def-rhs : Type-Stx Boolean Boolean -> Syntax
 (define (make-contract-def-rhs type flat? maker?)
-  (contract-def-property #'#f `#s(contract-def ,type ,flat? ,maker? untyped)))
+  (define contract-def `#s(contract-def ,type ,flat? ,maker? untyped))
+  (contract-def-property #'#f (λ () contract-def)))
+
+;; make-contract-def-rhs/from-typed : Id Boolean Boolean -> Syntax
+(define (make-contract-def-rhs/from-typed id flat? maker?)
+  (contract-def-property
+   #'#f
+   (λ () `#s(contract-def ,(or (cast-table-ref id) #'Any) ,flat? ,maker? typed))))
```

Comment:
And in the `Any` fallback version, it's still generating the correct contract, so the cast-table mutation is happening correctly, and the contract-generation pass gets the right contract. So it must be something else that's calling the thunk. So in the `Any` fallback version, this:
```racket
#lang racket
(require racket/sandbox)
(define ev (call-with-trusted-sandbox-configuration
             (λ () (make-module-evaluator "#lang typed/racket"))))
(ev '((cast (lambda ([s : String]) s) (Any -> Any)) 5))
```
Generates the correct error:
```
#<procedure:val>: contract violation
  expected: String
  given: 5
  in: the 1st argument of
      (-> String any)
  contract from: typed-world
  blaming: cast
   (assuming the contract is correct)
  at: eval:6.0
```
Where if it were getting the `Any`, it would raise a different error saying `Attempted to use a higher-order value passed as `Any` in untyped code: #<procedure:val>`. Since it's not that, we know that has gotten the correct type by the time the contract-generation pass happens.